SelfAndPeerEvaluations/utils.py

```diff
@@ -0,0 +1,439 @@
+import canvasapi
+from collections import defaultdict
+import csv
+from collections import deque
+import json
+import statistics
+import random
+
+
+class PartnerEvalQuizIndividualStats:
+    def __init__(self, name: str, id: int, files: list):
+        self.name = "".join(filter(str.isalnum, name))
+        self.id = str(id)
+        self.scores = defaultdict(list)
+
+        self.detailed_scores = []
+        self.qualitative_subjects = None
+
+        self.csv_file_path = (
+            f"./quiz_results/individual_student_reports/{self.name}.csv"
+        )
+        f = open(self.csv_file_path, "w")
+        self.writer = csv.writer(f)
+
+        self.__read_files(files)
+        self.__get_averages()
+        # transpose detailed scores
+        self.detailed_scores = [
+            [row[i] for row in self.detailed_scores]
+            for i in range(len(self.detailed_scores[0]))
+        ]
+        self.__format_csv_info()
+        self.writer.writerows(self.csv_file_info)
+        f.close()
+
+    def __format_csv_info(self) -> None:
+        """
+        formats self.csv_file_info for writing
+        """
+        self.csv_file_info = deque(self.detailed_scores)
+        self.csv_file_info.appendleft(["diff = self - partner"])
+        self.csv_file_info.appendleft(
+            ["INDIVIDUAL ASSIGNMENT GRADES\nAND CALCULATIONS"]
+        )
+        self.csv_file_info.extendleft([""] for _ in range(4))  # empty rows
+        self.csv_file_info.extendleft(self.__write_contribution_score())
+        self.csv_file_info.extendleft(self.__write_qualitative_score())
+        self.csv_file_info.appendleft([""])  # empty row
+        self.csv_file_info.appendleft(self.__write_final_score())
+        self.csv_file_info.appendleft([""])
+        self.csv_file_info.appendleft(["FINAL SCORE CALCULATIONS"])
+
+    def __give_default_scores(self, json_file: dict, row: list) -> None:
+        # default self&partner qualitative score is 1
```

Comment:
- Docstring needed
- Better hinting needed for ` json_file` and `row` if possible
-